modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/BaseUpgradeCheck.java

```diff
@@ -46,13 +54,86 @@ protected String addNewImports(String newContent) {
 		return newContent;
 	}
 
+	protected String addNewImportsJSP(String newContent, String... newImports) {
+		if (newImports.length == 0) {
+			return newContent;
+		}
+
+		Set<String> missingImports = new TreeSet<>();
+
+		Collections.addAll(missingImports, newImports);
+
+		Matcher directiveMatcher = _includesPattern.matcher(newContent);
+
+		if (directiveMatcher.find()) {
+			String includeDirective = directiveMatcher.group();
+
+			String[] includeDirectives = StringUtil.splitLines(
+				includeDirective);
+
+			String newHeaderPackage = createImportsPackageJsp(
+				missingImports, includeDirectives);
+
+			newContent = StringUtil.replaceFirst(
+				newContent,
+				com.liferay.petra.string.StringUtil.merge(
+					includeDirectives, "\n"),
+				newHeaderPackage);
+		}
+		else {
+			Matcher commentMatcher = _commentsPattern.matcher(newContent);
+
+			if (commentMatcher.find()) {
+				String comment = commentMatcher.group(1);
+
+				if (commentMatcher.start() > 1) {
+					return newContent;
+				}
+
+				String newMissingImports = createImportsPackageJsp(
+					missingImports, comment);
+
+				newContent = StringUtil.replaceFirst(
+					newContent, comment, newMissingImports);
+			}
+			else {
+				for (String missingImport : missingImports) {
+					newContent = StringBundler.concat(
+						"<%@ page import=\"", missingImport, "\" %>",
+						StringPool.NEW_LINE, newContent);
+				}
+			}
+		}
+
+		return newContent;
+	}
+
 	protected String afterFormat(
 		String fileName, String absolutePath, String content,
 		String newContent) {
 
 		return addNewImports(newContent);
 	}
 
+	protected String createImportsPackageJsp(
+		Set<String> missingImports, String... oldHeaderPackage) {
```

Comment:
@TamyrisTorres, what exactly do you mean with `oldHeaderPackage`?

I feel that it's a little inconsistent when we look to the parameter being passed to this method. Does it make sense to you?

Reply:
resolved